Add draft for august (#5)
nrc/tmit: august-19.md

# This month in TiKV: August 2019

Happy day! Welcome to the second edition of 'This Month in TiKV', covering August 2019. (Yes we know it's later than the first week of September, sorry!)

> The TiKV authors have been busy working on improving stability, fixing bugs, and laying out the foundations of TiKV 4.0. That's a lot to cover so lets get started!

Reviewer: ```suggestion
The TiKV authors have been busy working on improving stability, fixing bugs, and laying out the foundations of TiKV 4.0. That is a lot to cover, so let's get started!
```

## News

> This month our team made four TiKV releases! These minor releases include bugfixes and minor, backwards compatible features.

Reviewer: ```suggestion
This month our team made four small TiKV releases! These releases include bug fixes and minor, backwards compatible features.
```

You can review the changelogs here:

* [3.0.3](https://github.com/tikv/tikv/releases/tag/v3.0.3)
* [3.0.2](https://github.com/tikv/tikv/releases/tag/v3.0.2)
* [2.1.17](https://github.com/tikv/tikv/releases/tag/v2.1.17)
* [2.1.16](https://github.com/tikv/tikv/releases/tag/v2.1.16)

Upgrading? Please take note of these things:

* In **2.1.17 and 3.0.3** a new `config-check` option allows you to verify a configuration is correct before trying to start a node with it.
* In **3.0.2** some logs which require no manual intervention (TiKV resolves them automatically) were reduced to `INFO` level.
* In **2.1.16** `raw_scan` and `raw_batch_scan` now supports reverse.

> ## Reading material

Reviewer: ```suggestion
## Reading materials
```

> Here's some things our contributors have published over the last month:

Reviewer: ```suggestion
Here are some articles our contributors have published over the last month:
```

> * [@sunxiaoguang] published some details of the deployment including TiKV at Zhihu in [Lesson Learned from Queries over 1.3 Trillion Rows of Data Within Milliseconds of Response Time at Zhihu.com](https://pingcap.com/success-stories/lesson-learned-from-queries-over-1.3-trillion-rows-of-data-within-milliseconds-of-response-time-at-zhihu/), some highlights from article:

Reviewer: ```suggestion
* [@sunxiaoguang] published some details of the deployment including TiKV at Zhihu in [Lesson Learned from Queries over 1.3 Trillion Rows of Data Within Milliseconds of Response Time at Zhihu.com](https://pingcap.com/success-stories/lesson-learned-from-queries-over-1.3-trillion-rows-of-data-within-milliseconds-of-response-time-at-zhihu/). Some highlights from the article are:
```

* [@siddontang] published [Porting TiDB to ARM64 for Greater Flexibility](https://pingcap.com/blog/porting-tidb-to-arm64-for-greater-flexibility/) where he explores building and benchmarking TiKV and TiDB on ARM64 architectures.
* [@ethercflow] published a follow up article how [TCP Small Queues become a bottleneck for our test deployments on ARM64 architectures](https://pingcap.com/blog/how-tsq-becomes-a-performance-bottleneck-for-tikv-in-aws-arm-environment/).
* [@c4pt0r] published [(CN) A Talk about the New Features of TiDB: From Follower Read](https://pingcap.com/blog-cn/follower-read-the-new-features-of-tidb/) where he discusses follower read, a new feature we're planning for TiKV to be leveraged by query layers like TiDB.
* [@zhangjinpeng1987] published a new [(CN) Source Code Reading: Distributed Transactions](https://pingcap.com/blog-cn/tikv-source-code-reading-12/) discussing the transaction model powering TiKV.

## Roadmap

We've also began nailing down which features we want to include in a future TiKV 4.0 release. Here's a few of our hopes and dreams:

> (This is not an official list and may be subject to change)

> * Publish our RocksDB wrapper as `tirocks` on [crates.io](https://crates.io/crates/tirocks/).

Reviewer 2: seem we have not published it yet? only occupy the name

Author: Yes! That's why it's on the ROADMAP. :)

> * Abstract our engine so that we adapt to other storage technologies and offer more flexibility.

Reviewer 2: May using `start to abstrace` is better?

Reviewer: These are Roadmap items - it indicates something just started or to be started.  So I  think it is fine here.

* Begin adopting Joint Consensus, first to allow our clusters to more safely replace nodes in running deployments.

> * Further battle-hardening and improvements to the titan storage backend.

Reviewer: ```suggestion
* Further battle-hardening and improvements to the Titan storage backend.
```

* Support Quiescent Region, reducing the heartbeats of inactive ('cold') regions.
* Further coprocessor support and better documentation around coprocessor functionality, making it even easier to [make your first contribution to TiKV](https://pingcap.com/blog/adding-built-in-functions-to-tikv/).
* A 1.0.0 release of our Rust client, along with more improvements to our other clients.
* Reduce WAL usage to lower write pressure.

## Notable PRs

* In TiKV 2.1.17 gRPC was upgraded to fix a segfault causing abnormal exits [#5441](https://github.com/tikv/tikv/pull/5441).
* In TiKV 2.1.17 and 3.0.3 incorrect timestamps should no longer be reported from a Region [#5296](https://github.com/tikv/tikv/pull/5296).
* In TiKV 3.0.3 a possible request drop from ReadIndex when there is no leader was fixed [#5316](https://github.com/tikv/tikv/pull/5316).
* TiKV 2.1.16 now returns region errors when closing so the client will retry on other nodes [#4820](https://github.com/tikv/tikv/pull/4820).
* In TiKV 3.0.2 we fixed a bug in constraint checking during insertion when pessimistic transactions are enabled [#5128](https://.github.com/tikv/tikv/pull/5128).

> * In TiKV 3.0.2 we fixed the problem that TiKV lose some logs while panicking [#5174](https://github.com/tikv/tikv/pull/5174).

Reviewer: ```suggestion
* In TiKV 3.0.2 we fixed the problem that TiKV loses some logs while panicking [#5174](https://github.com/tikv/tikv/pull/5174).
```

## Current projects

Here's some of the things our contributors have been working on over the last month:

* [@sticnarf] and [@nrc] have been working on transaction support in our Rust Client.
* [@fullstop000] has been working to implement follower replication in Raft, for future use in TiKV.
* [@hicqu]has been leading the effort on supporting Joint Consensus in TiKV.
* Several members of the [@pingcap] team have been working on further shrinking and optimizing TiKV's build graph, resulting in faster compile times and less dependencies.
* [@zhouqiang-cl] and [@hoverbear] have been working on improving our release process and docker containers.
* [@dcalvin] and [@hoverbear] have been overhauling the TiKV.org documentation to better support our users.
* [@busyjay] and [@nrc] have been collaborating to land the final PRs to support using [Prost](https://github.com/danburkert/prost) library.
* [@nrc] has been working on Prost to help land some optimizations.
* [@overvenus] has been driving forward our new backup and restore process for TiKV.

> * [@MyonKeminta] has been leading on refining and hardening our experimental pessimistic transaction support.

Reviewer 3: I think it's @youjiali1995 who leads pessimistic transaction.

Author: Oh! Thank you!

* [@jiyingtk]
* [@eurekaka]
* [@jarifibrahim]
* [@ethan-daocloud]
* [@marsishandsome]
* [@you06]
* [@wujy-cs]
* [@928234269]
* [@lucklove]

If you'd like to get involved, we'd love to help you get started. You might be interested in tackling one of [these issues](https://github.com/tikv/tikv/issues?q=is%3Aopen+is%3Aissue+label%3A%22D%3A+Easy%22+label%3A%22S%3A+HelpWanted%22). If you don't know how to begin, please leave a comment and somebody will help you out. We're also very keen for people to contribute documentation, tests, optimizations, benchmarks, refactoring, or other useful things.

## This Week in TiDB

For more detailed and comprehensive information about TiDB and TiKV, we have weekly updates. The following cover August,

* [2019-08-04](https://pingcap.com/weekly/2019-08-05-tidb-weekly/)
* [2019-08-12](https://pingcap.com/weekly/2019-08-12-tidb-weekly/)
* [2019-08-19](https://pingcap.com/weekly/2019-08-19-tidb-weekly/)
* [2019-08-26](https://pingcap.com/weekly/2019-08-26-tidb-weekly/)

> [@sunxiaoguang]: https://github.com/

Reviewer 3: Seems the username part is missing

> [@sunxiaoguang]: https://github.com/

Reviewer: ```suggestion
[@sunxiaoguang]: https://github.com/sunxiaoguang/
```

* [@dcalvin] and [@hoverbear] have been overhauling the TiKV.org documentation to better support our users.
* [@busyjay] and [@nrc] have been collaborating to land the final PRs to support using [Prost](https://github.com/danburkert/prost) library.
* [@nrc] has been working on Prost to help land some optimizations.
* [@overvenus] has been driving forward our new backup and restore process for TiKV.
* [@MyonKeminta] has been leading on refining and hardening our experimental pessimistic transaction support.
* [@zhouqiang-cl]'s team at [@pingcap] has been working to improve our CI and automation to help reduce time-to-merge, expand testing, and improve automation.
* [@yiwu-arbug] and [Connor1996] are battle hardening Titan and incorporating feedback from some of our users.
* [@sunxiaoguang] has been battle testing TiKV on massive, China-scale workloads at Zhihu ([See his article!](https://pingcap.com/success-stories/lesson-learned-from-queries-over-1.3-trillion-rows-of-data-within-milliseconds-of-response-time-at-zhihu/))
* [@busyjay] and [@zhangjinpeng1987] have been evaluating bloom filters.
* [@5kbpers], [@brson] and [@aknuds1] have begun work on the engine abstraction project.

> * [MyonKeminta] has been driving a project to facilitate larger transactions.

Reviewer 4: ```suggestion
* [@MyonKeminta] has been driving a project to facilitate larger transactions.
```
